code/modules/vehicles/_hitbox.dm

```diff
@@ -46,6 +46,9 @@
 /obj/hitbox/Destroy(force)
 	if(!force) // only when the parent is deleted
 		return QDEL_HINT_LETMELIVE
+	for(var/mob/living/desant AS in tank_desants) // Lets clean up our riders before deleting.
+		if(HAS_TRAIT(desant, TRAIT_TANK_DESANT))
+			desant.remove_traits(list(TRAIT_TANK_DESANT, TRAIT_NOSUBMERGE), VEHICLE_TRAIT)
```

Comment:
So I spent quite a bit of time actually doing what you just said and I landed on this option for doing it this way for two reasons.

1. on_exited() has checks that will not allow it to actually run if the /living is still on top of the tank. Before qdel, this is still true, but after references might be cleared. I'm still fresh off the block though, so feel free to correct me on this one.

2. qdel and garbage collection already handles nulling variables and it's possibly even more performant to let the garbage collector work it's magic. The only thing it isn't doing is removing this trait from an unrelated /living entity.

I did make my own remove_desants() proc and just transfered all the code there, but it felt a bit ugly. Not only that, later down in the code it also checked again to see if there was a vehicle hitbox in our location, which wouldn't pass due to the former issue.

Reply:
1) this is not a case where performance is a concern
2) As you said it only handles itself: an example is that layers are broken BECAUSE that proc doesnt get called to clean up even with this proc
3) fair enough that it doesnt work, you may want to add a special proc that puts the actual desant removal seperately (ie instead of on exited you have a on exited (this has the checks) and a "remove_desant" or something proc that gets called by both these cases that actually handles the cleanup without any checks

Reply:
It's overall cleaner to do what you said, yeah. I'll go ahead with it.